### PR TITLE
WIP

### PR DESCRIPTION
**Pull request by yohanboniface**

### CODE REVIEW

**Comment by yohanboniface** [comment]
@davidbgk up for a pair review in the incoming days?

**Comment by davidbgk** [comment]
@yohanboniface on 27 or 28

File: usine.py [hunk 1]
```diff
@@ -1,23 +1,345 @@
...
+
+def get(remote, local):
+    bar = ProgressBar(prefix=f'{remote} => {local}')
+    func = client.sftp.getfo if hasattr(local, 'write') else client.sftp.get
+    func(remote, local, lambda done, total: bar.update(done=done, total=total))
+
+
+@contextmanager
+@formattable
+def sudo(set_home=True, preserve_env=True, user=None, login=True):
```
Comment (yohanboniface) [inline usine.py]: Update client.context `login` value according to `user` value.

File: usine.py [hunk 2]
```diff
@@ -1,23 +1,345 @@
...
+    prefix = 'sudo {set_home:B} {preserve_env:B} {user:V} {login:B}'
```
Comment (yohanboniface) [inline usine.py]: Can we use full word spec?

File: usine.py [hunk 3]
```diff
@@ -1,23 +1,345 @@
...
+@formattable
+def ls(path, all=True, human_readable=True, size=True, list=True):
+    return run('ls {all:B} {human_readable:B} {size:B} {list:S} {path}')
+
+
+def mv(src, dest):
+    return run(f'mv {src} {dest}')
+
+
+def put(local, remote, owner=None):
```
Comment (yohanboniface) [inline usine.py]: Can we use the ssh client instead, so we can take advantage of `with sudo/with cd` and such?

File: usine.py [hunk 4]
```diff
@@ -1,23 +1,345 @@
...
+class Config:
+
+    def __init__(self, value=None):
+        if isinstance(value, Config):
+            value = value.value
+        super().__setattr__('value', value or {})
+
+    def get(self, key, default=None):
+        try:
+            return Config(self.value.get(key, default))
```
Comment (yohanboniface) [inline usine.py:43]: Try a simpler version, without each returning a new instance

File: usine.py [hunk 5]
```diff
@@ -1,23 +1,349 @@
...
+
+            # if there's a field, output it
+            if name:
+                # given the name, find the object it references
+                #  and the argument it came from
+                obj, _ = self.get_field(name, args, kwargs)
+                obj = self.convert_field(obj, conversion)
+
+                value = ''
+                if spec == 'bool':
```
Comment (davidbgk) [inline usine.py:128]: I wonder if we should extract that part into a dedicated method in order to ease subclassing (additional specs).

File: usine.py [hunk 6]
```diff
@@ -1,23 +1,349 @@
...
+            if idx < len(args):
+                client.context[name] = args[idx]
+            else:
+                client.context[name] = kwargs.get(name, param.default)
+        return func(*args, **kwargs)
+
+    return wrapper
+
+
+class Status:
```
Comment (yohanboniface) [inline usine.py:163]: Shall we use/inherit from [subprocess.CompletedProcess](https://docs.python.org/3/library/subprocess.html#subprocess.CompletedProcess) instead?

File: usine.py [hunk 7]
```diff
@@ -1,23 +1,349 @@
...
+
+def get(remote, local):
+    bar = ProgressBar(prefix=f'{remote} => {local}')
+    func = client.sftp.getfo if hasattr(local, 'write') else client.sftp.get
+    func(remote, local, lambda done, total: bar.update(done=done, total=total))
+
+
+@contextmanager
+@formattable
+def sudo(set_home=True, preserve_env=True, user=None, login=None):
```
Comment (yohanboniface) [inline usine.py]: Could we also use it as a decorator?
```
@cli
@sudo
def my_cmd():
   ...
```

File: usine.py [hunk 8]
```diff
@@ -1,23 +1,362 @@
...
+    return run('chown {recursive:bool} {mode} {path}')
+
+
+@formattable
+def ls(path, all=True, human_readable=True, size=True, list=True):
+    return run('ls {all:bool} {human_readable:bool} {size:bool} {list:initial}'
+               ' {path}')
+
+
+def mv(src, dest):
```
Comment (davidbgk) [inline usine.py:363]: Is it still relevant now we have `cp`?

File: usine.py (hunk 7 above)
Comment (davidbgk) [reply usine.py]: To be consistent we would have to allow that for `cd` and `env` too?

File: usine.py (hunk 8 above)
Comment (yohanboniface) [reply usine.py:363]: Good catch (wrong copy-paste)!

**Review by davidbgk — APPROVED** [review]
r+remove `cp` as discussed.

**Comment by yohanboniface** [comment]
> remove cp as discussed

Oops, I don't recall what's wrong with cp :p

**Comment by davidbgk** [comment]
Meh, remove `mv` I meant…

**Comment by yohanboniface** [comment]
euh beh, why would we remove "mv"?

**Comment by davidbgk** [comment]
I cant remember the discussion associated :s

**Pull request merged**